php/class-wp-https-detection.php

```diff
@@ -0,0 +1,144 @@
+	const REQUEST_TOKEN_QUERY_ARG = 'https_detection_token';
+
+	/**
+	 * Inits the class.
+	 */
+	public function init() {
+		add_action( 'wp', array( $this, 'schedule_cron' ) );
+		add_action( self::CRON_HOOK, array( $this, 'update_option_https_support' ) );
+		add_filter( 'cron_request', array( $this, 'ensure_http_if_sslverify' ), PHP_INT_MAX );
+	}
+
+	/**
+	 * Schedules a cron event to check for HTTPS support.
+	 */
+	public function schedule_cron() {
+		if ( ! wp_next_scheduled( self::CRON_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::CRON_HOOK );
+		}
+	}
+
+	/**
+	 * Makes a request to find whether HTTPS is supported, and stores the result in an option.
+	 */
+	public function update_option_https_support() {
+		update_option( self::HTTPS_SUPPORT_OPTION_NAME, $this->is_https_supported() );
+	}
+
+	/**
+	 * Makes a request to the home URL to determine whether HTTPS is supported.
+	 *
+	 * Returns true if the URL has the scheme of HTTPS, and it has the correct token.
+	 * The token is only to verify that the request is for the same site.
+	 * Also, in wp_remote_get(), 'sslverify' is true by default.
+	 *
+	 * @return boolean Whether HTTPS is supported.
+	 */
+	public function is_https_supported() {
+		$request = wp_remote_request(
+			add_query_arg(
+				self::REQUEST_TOKEN_QUERY_ARG,
+				$this->get_token(),
+				home_url( '/', 'http' )
+			),
+			array(
+				'timeout' => 10,
+				'headers' => array(
+					'Cache-Control' => 'no-cache',
+				),
+			)
+		);
+
+		if ( is_wp_error( $request ) || ! method_exists( $request['http_response'], 'get_response_object' ) ) {
+			return false;
+		}
+
+		$response   = $request['http_response']->get_response_object();
+		$parsed_url = wp_parse_url( $response->url );
+		if ( ! $response->success || ! isset( $parsed_url['query'] ) ) {
+			return false;
+		}
+
+		$query_args = wp_parse_args( $parsed_url['query'] );
+		$is_https   = (
+			isset( $parsed_url['scheme'], $query_args[ self::REQUEST_TOKEN_QUERY_ARG ] )
+			&&
+			'https' === $parsed_url['scheme']
+			&&
+			$this->get_token() === $query_args[ self::REQUEST_TOKEN_QUERY_ARG ]
+		);
+
+		return $is_https;
+	}
+
+	/**
+	 * Gets the token, based on the secret.
+	 *
+	 * This token is not intended for security, as much as ensuring that the request is for this site.
+	 * By calling wp_hash(), it uses this site's salts, which should be different from other sites.
+	 *
+	 * @return string $token The token for the request.
+	 */
+	public function get_token() {
+		return wp_hash( self::REQUEST_SECRET );
+	}
+
+	/**
+	 * If the 'cron_request' arguments include an HTTPS URL, this ensures sslverify is false.
+	 *
+	 * Prevents an issue if HTTPS breaks,
+	 * where there would be a failed attempt to verify HTTPS.
+	 *
+	 * @param array $request The cron request arguments.
+	 * @return array $request The filtered cron request arguments.
+	 */
+	public function ensure_http_if_sslverify( $request ) {
+		if ( 0 === strpos( $request['url'], 'https' ) ) {
```

Comment:
I suggest replacing `https` with `https://` for good measure.

Reply:
Sure, [it's now](https://github.com/xwp/pwa-wp/pull/16/files#diff-1d91a59d72444acecd2a72e8baa63c2cR119) `https://`.